Add AMD GPU support (#1546)
huggingface/optimum: docs/source/onnxruntime/usage_guides/amdgpu.mdx

> # Accelerated inference on AMD GPUs

Reviewer: ```suggestion
# Accelerated inference on AMD GPUs supported by ROCm
```

> By default, ONNX Runtime runs inference on CPU devices. However, it is possible to place supported operations on an AMD Instinct GPU, while leaving any unsupported ones on CPU. In most cases, this allows costly operations to be placed on GPU and significantly accelerate inference.

Author: Done

This guide will show you how to run inference on the `ROCMExecutionProvider` execution provider that ONNX Runtime supports for AMD GPUs:

## Installation

> #### 1. ROCM Installation (V 5.7.X)

Reviewer: ```suggestion
#### 1. ROCm Installation (V 5.7.X)
```
`(V 5.7.X)` means that ROCm 5.7 is a requirement? Can we put that rather in a sentence?

Author: The branch AMD team shared for onnxruntime was on ROCm 5.7. So this has been tested on the ROCm5.7. I could mention that the following instructions are torun the ROCMEP on AMD gpu with ROCm 5.7 installed.

To install ROCM, please follow the [ROCM installation guide](https://rocm.docs.amd.com/en/latest/deploy/linux/index.html).

> #### 2. Torch Installation with ROCM Support

Reviewer: ```suggestion
#### 2. PyTorch installation with ROCm support
```

> 

Reviewer: ```suggestion

Optimum ONNX Runtime integration relies on some functionalities of Transformers that require PyTorch. For now, we recommend to use Pytorch compiled against RoCm 5.7, that can be installed following [PyTorch installation guide](https://pytorch.org/get-started/locally/): 
```

Author: Added

```bash
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/rocm5.7
```

> > **Note1:** Exporting with SDPA may encounter issues with torch versions below 2.1.0. To ensure smooth operation, consider upgrading to torch 2.1.0 or newer.

Reviewer: ```suggestion
```

The suggestion is PyTorch nightly in the above. Also fixed in https://github.com/huggingface/optimum/pull/1547

> > **Note2:** For docker installation, the following base image is recommended: rocm/pytorch:rocm5.7_ubuntu22.04_py3.10_pytorch_2.0.1

Reviewer: ```suggestion
<Tip>
For docker installation, the following base image is recommended: `rocm/pytorch:rocm5.7_ubuntu22.04_py3.10_pytorch_2.0.1`
</Tip>
```

> ### 3. ONNXRuntime Installation with ROCM Execution Provider

Reviewer: ```suggestion
### 3. ONNX Runtime installation with ROCM Execution Provider
```

```bash
# pre-requisites
pip install -U pip
pip install cmake onnx
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

# Install ONNXRuntime from source
git clone --recursive  https://github.com/ROCmSoftwarePlatform/onnxruntime.git
git checkout rocm5.7_internal_testing_eigen-3.4.zip_hash
cd onnxruntime

./build.sh --config Release --build_wheel --update --build --parallel --cmake_extra_defines ONNXRUNTIME_VERSION=$(cat ./VERSION_NUMBER) --use_rocm --rocm_home=/opt/rocm
pip install build/Linux/Release/dist/*
```

> > **Note:** To avoid conflicts between `onnxruntime` and `onnxruntime-rocm`, make sure the package `onnxruntime` is not installed by running `pip uninstall onnxruntime` prior to installing `onnxruntime-rocm`.

Reviewer: ```suggestion
<Tip>
To avoid conflicts between `onnxruntime` and `onnxruntime-rocm`, make sure the package `onnxruntime` is not installed by running `pip uninstall onnxruntime` prior to installing `onnxruntime-rocm`.
</Tip>
```

> ### Checking the ROCM installation is successful

Reviewer: ```suggestion
### Checking the ROCm installation is successful
```

Before going further, run the following sample code to check whether the install was successful:

```python
>>> from optimum.onnxruntime import ORTModelForSequenceClassification
>>> from transformers import AutoTokenizer

>>> ort_model = ORTModelForSequenceClassification.from_pretrained(
...   "philschmid/tiny-bert-sst2-distilled",
...   export=True,
...   provider="ROCMExecutionProvider",
... )

>>> tokenizer = AutoTokenizer.from_pretrained("philschmid/tiny-bert-sst2-distilled")
>>> inputs = tokenizer("expectations were low, actual enjoyment was high", return_tensors="pt", padding=True)

>>> outputs = ort_model(**inputs)
>>> assert ort_model.providers == ["ROCMExecutionProvider", "CPUExecutionProvider"]
```

In case this code runs gracefully, congratulations, the installation is successfull! If you encounter the following error or similar,

```
ValueError: Asked to use ROCMExecutionProvider as an ONNX Runtime execution provider, but the available execution providers are ['CPUExecutionProvider'].
```

then something is wrong with the ROCM or ONNX Runtime installation.

> ### Use ROCM execution provider with ORT models

Reviewer: ```suggestion
### Use ROCM Execution Provider with ORT models
```

# Accelerated inference on AMD GPUs supported by ROCm

By default, ONNX Runtime runs inference on CPU devices. However, it is possible to place supported operations on an AMD Instinct GPU, while leaving any unsupported ones on CPU. In most cases, this allows costly operations to be placed on GPU and significantly accelerate inference.

> Our testing primarily involved AMD Instinct GPUs, and for specific GPU compatibility, refer to the support matrix available at AMD ROCm GPU OS support matrix."

Reviewer: ```suggestion
Our testing involved AMD Instinct GPUs, and for specific GPU compatibility, please refer to the official support list of GPUs available [here](https://rocm.docs.amd.com/en/latest/release/gpu_os_support.html).
```
